esp-hal-embassy/src/time_driver.rs

```diff
@@ -121,7 +122,8 @@ impl EmbassyTimer {
     fn arm(timer: &mut Timer, timestamp: u64) {
         let now = current_time().duration_since_epoch();
         let ts = timestamp.micros();
-        let timeout = if ts > now { ts - now } else { now };
+        // if the TS is already in the past make the timer fire immediately
+        let timeout = if ts > now { ts - now } else { 0.micros() };
```

Comment:
this is the actual fix - the other changes are just "cosmetics"

Reply:
Do we have a test for the issue in place? I'm not sure this is the optimal way for us to implement a time driver, and I'd like to know if changing things would break it.

I think all of our hardware triggers an alarm if we set an old timestamp. This means we shouldn't need to check if `timestamp` is past, we shouldn't reset the underlying timer's counters, we should simply handle wraparound and set the compare value.

Reply:
Idk if we finally got a test - when this got fixed Juraj was working on tests for the timer (apparently from the conversation here - can't remember t.b.h.)

> I think all of our hardware triggers an alarm if we set an old timestamp

IIRC there were some problems around that - maybe @jessebraham remembers? (I think you had some fun with timers and interrupts in the past - but maybe other issues or maybe I am just mixing things up)

Reply:
I seem to recall encountering problems with that but can't remember any specifics, will try to look at some old PRs and see if I can remember. Will update if I do end up remembering anything.

Reply:
Regardless if we know that a timestamp is already past, we can avoid using the timer completely and ask embassy to do a re-poll for us, without the additional latency of an interrupt (that may already have fired once). I am looking into these things, but I'd like to be confident in knowing that I'm not reintroducing old bugs.